cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/PipeTransformDetail.java

```diff
@@ -49,8 +50,15 @@ public void process(KeyValue<String, Object> value) {
       } else {
         transformation.transform(value, emitter);
       }
+    } catch (StageFailureException e) {
+      /* Another stage has already failed, just throw the exception as-is */
+      throw e;
     } catch (Exception e) {
-      throw new RuntimeException(e);
+      /* Create StageFailureException to save the Stage information */
+      throw new StageFailureException(
+        String.format("Failed to execute pipeline stage '%s' with the error: %s. Please review your pipeline " +
+                        "configuration and check the system logs for more details.", stageName,
+                      Throwables.getRootCause(e).getMessage()), Throwables.getRootCause(e));
```

Comment:
Get the root cause once and call getMessage() on it instead of getting the root cause twice.